### PR TITLE
Fix back compat full config versioning

### PR DESCRIPTION
**Pull request by tyler-cai-microsoft**
[AB#4657](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/4657)

The back-compat with loader pipeline broke because we were getting `0.0.0-<buildId>-test` from our pkg version, and it was expecting either `2.0.0-internal.x.y.z` or `2.0.0-dev.x.y.z`. The pipeline should be fine whether or not -test is appended to the end of the version, i.e `2.0.0-dev.x.y.z-test` 

Not sure what the right fix is here, but these are some options. I'd like to know if anyone knows how to test this.

### CODE REVIEW

**Review by alexvy86 — COMMENTED** [review]
Seems like it'll do the trick but I'd wait for @tylerbutler 's feedback.

**Comment by tylerbutler** [comment]
I need a lot more context for this change. Did we suddenly start publishing packages with a different version scheme?

**Comment by tyler-cai-microsoft** [comment]
> I need a lot more context for this change. Did we suddenly start publishing packages with a different version scheme?

Specifically the test branches have pkgVersion` 0.0.0-<buildId>-test` instead of `2.0.0-dev.x.y.z-test`. I was expecting that the test branch pkg versions to be `2.0.0-dev.x.y.z-test`. Otherwise the back compat tests will not run correctly for test/<name> pipelines

**Comment by tylerbutler** [comment]
> > I need a lot more context for this change. Did we suddenly start publishing packages with a different version scheme?
> 
> Specifically the test branches have pkgVersion` 0.0.0-<buildId>-test` instead of `2.0.0-dev.x.y.z-test`. I was expecting that the test branch pkg versions to be `2.0.0-dev.x.y.z-test`. Otherwise the back compat tests will not run correctly for test/ pipelines

The risk with the change is that we might accidentally publish a package with invalid dependencies. These tasks are meant to catch that:

https://github.com/microsoft/FluidFramework/blob/91b87c997784c9fc23e304e575ae38b3ed590ad7/tools/pipelines/templates/include-set-package-version.yml#L120-L146

If we can legitimately publish versions of the form 2.0.0-internal.x.y.z-test, then I think these last-minute checks need to be adjusted.

**Comment by tyler-cai-microsoft** [comment]
> > > I need a lot more context for this change. Did we suddenly start publishing packages with a different version scheme?
> > 
> > 
> > Specifically the test branches have pkgVersion` 0.0.0-<buildId>-test` instead of `2.0.0-dev.x.y.z-test`. I was expecting that the test branch pkg versions to be `2.0.0-dev.x.y.z-test`. Otherwise the back compat tests will not run correctly for test/ pipelines
> 
> The risk with the change is that we might accidentally publish a package with invalid dependencies. These tasks are meant to catch that:
> 
> https://github.com/microsoft/FluidFramework/blob/91b87c997784c9fc23e304e575ae38b3ed590ad7/tools/pipelines/templates/include-set-package-version.yml#L120-L146
> 
> If we can legitimately publish versions of the form 2.0.0-internal.x.y.z-test, then I think these last-minute checks need to be adjusted.

I'm going to try to avoid this problem altogether

File: packages/test/test-version-utils/src/compatConfig.ts [hunk 1]
```diff
@@ -154,14 +154,23 @@ const genFullBackCompatConfig = (): CompatConfig[] => {
...
+	if (
+		codeVersion !== undefined &&
+		(codeVersion.startsWith("2.0.0-internal.") || codeVersion.startsWith("2.0.0-dev."))
+	) {
+		version = codeVersion;
+	}
+	if (version.startsWith("2.0.0-internal.")) {
+		semverInternal = version.split("internal.")[1];
+	} else if (version.startsWith("2.0.0-dev.")) {
+		semverInternal = version.split("dev.")[1];
```
Comment (tylerbutler) [inline packages/test/test-version-utils/src/compatConfig.ts:169]: I worry about this code. Can we use the version-tools package for some of this? It has functions to decode/encode internal versions, validate them, etc, plus tests. I think with those APIs, this code looks more like this:

```ts
const [publicVersion, semverInternal, prereleaseIdentifier] = fromInternalScheme(codeVersion, true, true);
```

Comment (tylerbutler) [reply packages/test/test-version-utils/src/compatConfig.ts:169]: Relevant source: https://github.com/microsoft/FluidFramework/blob/547789bec19e64a7df2a04bb251c7c8939f6db8a/build-tools/packages/version-tools/src/internalVersionScheme.ts#L56

File: packages/test/test-version-utils/src/compatConfig.ts [hunk 2]
```diff
@@ -153,22 +154,21 @@ const genFullBackCompatConfig = (): CompatConfig[] => {
...
-		// This will need to be updated once we move beyond 2.0.0-internal.x.y.z
-		throw new Error(
-			`Unexpected back compat scenario! Expecting package versions to just be 2.0.0-internal.x.y.z, but got ${pkgVersion}.`,
-		);
+	let version: string = pkgVersion;
+	// This grabs the code version to find the backwards compatible options.
+	const codeVersion = process.env.SETVERSION_CODEVERSION;
+	if (
+		codeVersion !== undefined &&
+		(codeVersion.startsWith("2.0.0-internal.") || codeVersion.startsWith("2.0.0-dev."))
```
Comment (tylerbutler) [inline packages/test/test-version-utils/src/compatConfig.ts:162]: You should use isInternalVersionScheme for this. Any time you're checking the version string yourself, it's better to use the APIs.

File: packages/test/test-version-utils/src/compatConfig.ts [hunk 3]
```diff
@@ -153,22 +154,21 @@ const genFullBackCompatConfig = (): CompatConfig[] => {
...
+	) {
+		version = codeVersion;
 	}
 
+	const [_, semverInternal, __] = fromInternalScheme(version, true, true);
```
Comment (tylerbutler) [inline packages/test/test-version-utils/src/compatConfig.ts:167]:
```suggestion
	const [, semverInternal] = fromInternalScheme(version, true, true);
```

**Review by tylerbutler — APPROVED** [review]

**Pull request merged**